pyprophet/export_parquet.py

```diff
@@ -0,0 +1,562 @@
+    # we cannot convert NAs to integers
+    df[list(id_cols_datatype.keys())] = df[list(id_cols_datatype.keys())].fillna(-1)
+    df = df.astype(id_cols_datatype)
+
+    #############################
+    ## Other Int64 Cols
+
+    # Get dict of cols to assert datatype
+    id_cols_datatype = {col:'int64' for col in df.columns if re.search('RANK', col)}
+    # We need to fill na IDs with a -1 to force conversion to int64, because
+    # we cannot convert NAs to integers
+    df[list(id_cols_datatype.keys())] = df[list(id_cols_datatype.keys())].fillna(-1)
+    df = df.astype(id_cols_datatype)
+
+    return df
+
+def read_precursor_feature_data(con, columnsToSelect, prec_ids, run_id, outfile):
+        # Feature Data
+        feature_query = '''
+        SELECT {0}
+        FROM (SELECT * FROM FEATURE WHERE PRECURSOR_ID IN ({1}) AND RUN_ID = {2}) AS FEATURE
+        LEFT JOIN (SELECT * FROM RUN WHERE ID = {2}) AS RUN ON FEATURE.RUN_ID = RUN.ID
+        LEFT JOIN FEATURE_MS1 ON FEATURE.ID = FEATURE_MS1.FEATURE_ID
+        LEFT JOIN FEATURE_MS2 ON FEATURE.ID = FEATURE_MS2.FEATURE_ID
+        LEFT JOIN SCORE_MS2 ON FEATURE.ID = SCORE_MS2.FEATURE_ID
+        '''.format(','.join([col.strip() for col in columnsToSelect.split(',') if re.search('^FEATURE|^RUN|^SCORE_MS2', col.strip()) and not re.search('^FEATURE_TRANSITION', col.strip())] + ['FEATURE.PRECURSOR_ID']), ','.join(prec_ids), run_id)
+
+        # Read into Pandas Dataframe
+        df_feature = pd.read_sql(feature_query, con)
+
+        # Precursor Library Data
+        ## Check if Gene tables are present
+        if check_sqlite_table(con, "GENE"):
+            gene_table_joins = '''
+            LEFT JOIN PEPTIDE_GENE_MAPPING ON PEPTIDE.ID = PEPTIDE_GENE_MAPPING.PEPTIDE_ID
+            LEFT JOIN GENE ON PEPTIDE_GENE_MAPPING.GENE_ID = GENE.ID
+            '''
+        else:
+            gene_table_joins = ''
+
+        precursor_query = f'''
+        SELECT {','.join([col.strip() for col in columnsToSelect.split(',') if re.search('^PRECURSOR|^PRECURSOR_PEPTIDE_MAPPING|^PEPTIDE|^PEPTIDE_PROTEIN_MAPPING|^PROTEIN', col.strip())])}
+        FROM (SELECT * FROM PRECURSOR WHERE ID in ({','.join(prec_ids)})) AS PRECURSOR
+        LEFT JOIN PRECURSOR_PEPTIDE_MAPPING ON PRECURSOR.ID = PRECURSOR_PEPTIDE_MAPPING.PRECURSOR_ID
+        LEFT JOIN PEPTIDE ON PRECURSOR_PEPTIDE_MAPPING.PEPTIDE_ID = PEPTIDE.ID
+        LEFT JOIN PEPTIDE_PROTEIN_MAPPING ON PEPTIDE.ID = PEPTIDE_PROTEIN_MAPPING.PEPTIDE_ID
+        LEFT JOIN PROTEIN ON PEPTIDE_PROTEIN_MAPPING.PROTEIN_ID = PROTEIN.ID
+        {gene_table_joins}
+        '''
+        # Read into Pandas Dataframe
+        df_prec = pd.read_sql(precursor_query, con)
+
+        # Merge Feature and Precursor Tables
+        df_tmp = pd.merge(df_feature, df_prec, how='outer', on=['PRECURSOR_ID'])
+        # Fill in run for cases where precursors have no features identified
+        df_tmp['FEATURE.RUN_ID'] = df_tmp['FEATURE.RUN_ID'].dropna().unique()[0]
```

Comment:
@singjc I don't understand the purpose of this line? If there are no features identified shouldn't it not be associated with any run?